File: Kattis/NAC 2022/GCD_Harmony.py
import sys
input = sys.stdin.readline
from math import sqrt, gcd, inf
from collections import defaultdict
def multiple_ints():
  return map(int, input().strip().split())

def solve():  
  n = int(input())
  vals = [0 for i in range(n)]
  adj = [[] for i in range(n)]
  primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 
    53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
  cap = n * 2
  dp = []
  dp_prime_condense = [defaultdict(lambda: cap + 1) for i in range(n)]

  def factorize(num):
    factors = []
    for p in primes:
      if num == 1:
        break
      elif num % p == 0:
        factors.append(p)
        while num % p == 0:
          num //= p
    if num > 100:
      return []
    if num > 1:
      factors.append(p)
    
    return factors
  

  precomp = {}
  valid = []
  for i in range(2, cap+1):
    f = factorize(i)
    if f:
      precomp[i] = f
      valid.append(i)
  
  for i in range(n):
    dp.append({})
    for v in valid:
      if vals[i] == v:
        dp[i][v] = 0
      else:
        dp[i][v] = v
  #print(dp[0], "HERE ASA SAS")
  #dp where dp[i][j] is the cost to make node i gcd harmonic with all of its children given that node i has value j

  #ans will never exceed n * 2, even in degenerate prime cases

  #can optimize by doing coordinate compression and only including valid precomp values? May be slower due to dict lookup thouugh
  for i in range(n):
    v = int(input())
    vals[i] = v
    dp[i][v] = 0
    dp[i][0] = cap + 1
    dp[i][1] = cap + 1
  for i in range(n-1):
    a,b = multiple_ints()
    a -= 1
    b -= 1
    adj[b].append(a)
    adj[a].append(b)
  
  def dfs(curr, par):
    #print("HERE", curr, par)
    for child in adj[curr]:
      if child == par:
        continue
      dfs(child, curr)

      for val in valid:
        #if we went to set curr to val, we need to transition from any value in child that is GCD harmonic with val
        #this is just any value that shares a prime factor
        #condense all prime factors into one
        #we want to choose the prime factor w/ minimum cost
        child_cost = cap + 1
        for factor in precomp[val]:
          child_cost = min(child_cost, dp_prime_condense[child][factor])
        #print(child_cost, curr, child, val)
        dp[curr][val] += child_cost
      
      #print(dp[curr], curr, "HERE")
      #condense all values in DP
    for val in valid:
      for factor in precomp[val]:
        dp_prime_condense[curr][factor] = min(dp_prime_condense[curr][factor], dp[curr][val])
    #print("FINISHED", dp_prime_condense[curr])
    #print(dp[curr], "HERE", curr)
  dfs(0, -1)
  

  print(min(dp[0].values()))
  #print(dp)

solve()

File: Kattis/NAC 2022/GCD_Harmony.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> adj;
vector<int> vals;
vector<vector<int>> precomp;
vector<map<int, int>> dp_prime_condense;
vector<vector<int>> dp;
int cap = 2 * 5e3;
const std::vector<int> primes = {
        2, 3, 5, 7, 11, 13, 17, 19, 23, 29,
        31, 37, 41, 43, 47, 53, 59, 61, 67,
        71, 73, 79, 83, 89, 97
    };
    

vector<int> factorize(int num){
  vector<int> factors;
  for(int prime : primes){
    if(num == 1){
      break;
    }
    if(num % prime == 0){
      factors.push_back(prime);
      while(num % prime == 0){
        num /= prime;
      }
    }
  }

  return factors;
}

void dfs(int curr, int par){
  for(int child : adj[curr]){
    if(child == par){
      continue;
    }
    
    dfs(child, curr);

    for(int val = 2; val <= cap; val++){
      int child_cost = cap + 1;
      for(int fac : precomp[val]){
        child_cost = min(child_cost, dp_prime_condense[child][fac]);
      }
      dp[curr][val] += child_cost;
    }
  }
  for(int prime : primes){
    int prime_cost = cap + 1;
    for(int v = prime; v <= cap; v += prime){
      prime_cost = min(prime_cost, dp[curr][v]);
    }
    dp_prime_condense[curr][prime] = prime_cost;
  }
}

void solve(){
  int n; cin >> n;
  vals.assign(n, 0);
  adj.assign(n, vector<int>());
  dp.assign(n, vector<int>(cap+1, cap+1));
  dp_prime_condense.assign(n, map<int, int>());
  precomp.assign(cap+1, vector<int>());


  for(int i = 0; i < n; i++){
    cin >> vals[i];
  }

  for(int i = 0; i < n - 1;i++){
    int a; int b; cin >> a >> b;
    a--; b--;
    adj[a].push_back(b);
    adj[b].push_back(a);
  }
  

  for(int i = 0; i < n; i++){
    for(int val = 2; val <= cap; val++){
      if(val == vals[i]){
        dp[i][val] = 0;
      }
      else{
        dp[i][val] = val;
      }
    }
  }
  
  for(int i = 2; i <= cap; i++){
    precomp[i] = factorize(i);
  }

  dfs(0, -1);
  int ans = *min_element(dp[0].begin(), dp[0].end());
  cout << ans;
  //cout << " H E R E";
}

int main(){
  solve();
  return 0;
}